### PR TITLE
Add parallel execution to assist.

### PR DESCRIPTION
**Pull request by jakule**
Execute Assist commands in parallel. 

Closes https://github.com/gravitational/teleport.e/issues/1468

### CODE REVIEW

**Comment by jakule** [comment]
@timothyb89 @xacrimon Friendly ping

File: lib/web/command.go [hunk 1]
```diff
@@ -185,78 +186,106 @@ func (h *Handler) executeCommand(
...
 	h.log.Debugf("Found %d hosts to run Assist command %q on.", len(hosts), req.Command)
 
-	for _, host := range hosts {
-		err := func() error {
-			sessionData, err := h.generateCommandSession(&host, req.Login, clusterName, sessionCtx.cfg.User)
-			if err != nil {
-				h.log.WithError(err).Debug("Unable to generate new ssh session.")
-				return trace.Wrap(err)
-			}
+	runCmd := func(host *hostInfo) error {
```
Comment (xacrimon) [inline lib/web/command.go:189]: can we move this into a seperate function at all? `executeCommand` is already pretty unwieldy to look at.

File: lib/web/command.go [hunk 2]
```diff
@@ -185,78 +186,106 @@ func (h *Handler) executeCommand(
...
+		h.log.Debugf("New command request for server=%s, id=%v, login=%s, sid=%s, websid=%s.",
+			host.hostName, host.id, req.Login, sessionData.ID, sessionCtx.GetSessionID())
+
+		commandHandlerConfig := CommandHandlerConfig{
+			SessionCtx:         sessionCtx,
+			AuthProvider:       clt,
+			SessionData:        sessionData,
+			KeepAliveInterval:  netConfig.GetKeepAliveInterval(),
+			ProxyHostPort:      h.ProxyHostPort(),
+			InteractiveCommand: strings.Split(req.Command, " "),
```
Comment (xacrimon) [inline lib/web/command.go:205]: should we do this outside of the function that gets loop-called for every target node?

File: lib/web/command.go [hunk 3]
```diff
@@ -185,78 +186,106 @@ func (h *Handler) executeCommand(
...
+	runCommands(hosts, runCmd, h.log)
+
 	return nil, nil
 }
 
+// runCommands runs the given command on the given hosts.
+func runCommands(hosts []hostInfo, runCmd func(host *hostInfo) error, log logrus.FieldLogger) {
+	// Create a synchronization channel to limit the number of concurrent commands.
+	// The maximum number of concurrent commands is 30 - it is arbitrary.
+	syncChan := make(chan struct{}, 30)
```
Comment (xacrimon) [inline lib/web/command.go:260]: can/should we put this in a configuration resource somewhere? I can imagine someone wanting to play with/adjust this.

File: lib/web/command.go [hunk 4]
```diff
@@ -185,78 +186,106 @@ func (h *Handler) executeCommand(
...
+	// WaiteGroup to wait for all commands to finish.
+	wg := sync.WaitGroup{}
+
+	for _, host := range hosts {
+		host := host
+
+		wg.Add(1)
```
Comment (xacrimon) [inline lib/web/command.go:267]:
```suggestion
		host := host
		wg.Add(1)
```

File: lib/web/command.go (hunk 3 above)
Comment (xacrimon) [inline lib/web/command.go:260]: should this use `sync/semaphore`, seems like a redneck semaphore that has channel overheads in terms of space and whatnot, anyhow.

File: lib/web/command.go (hunk 1 above)
Comment (jakule) [reply lib/web/command.go:189]: I've tried, but this function uses 11 variables from the outer scope. I've tried to move them to a new struct, but IMO, at that point, it's not worth it. If you have a strong opinion about that I can do it.

File: lib/web/command.go (hunk 3 above)
Comment (jakule) [reply lib/web/command.go:260]: I was also thinking about it, but Go has divided opinions about semaphores in general. I remember a proposal to remove semaphores in Go 2, which had many examples of why you should not use than. Now even stdlib recommends "kind of" against them https://pkg.go.dev/sync#Cond

File: lib/web/command.go (hunk 2 above)
Comment (jakule) [reply lib/web/command.go:205]: Changed

File: lib/web/command.go (hunk 1 above)
Comment (xacrimon) [reply lib/web/command.go:191]: 👍 fine then

File: lib/web/command.go (hunk 3 above)
Comment (xacrimon) [reply lib/web/command.go:262]: that's uh, certainly a way to design a language. anyhow, in that case, fine by me.

**Review by xacrimon — APPROVED** [review]
lgtm but if we could get @justinas  to chime in on the config problem, that'd be nice

**Review by timothyb89 — APPROVED** [review]
This looks reasonable to me, I don't have strong opinions w.r.t. configuration.

File: lib/web/command.go (hunk 3 above)
Comment (justinas) [reply lib/web/command.go:262]: No concerns, we'll likely leave it as default on Cloud.

Comment (jakule) [reply lib/web/command.go:265]: @xacrimon  I created a new issue to address this comment, as I don't want to introduce more changes after people have already reviewed it. https://github.com/gravitational/teleport.e/issues/1516

Comment (xacrimon) [reply lib/web/command.go:265]: splendid

**Pull request merged**